src/MPI.jl

```diff
@@ -103,6 +103,10 @@ function __init__()
     end
 
     @static if Sys.isunix()
+        # Cray MPICH apparently requires you to manually load a separate GPU Transport Layer (GTL) if you want GPU-aware MPI.
+        if MPIPreferences.binary == "system" && MPIPreferences.abi == "MPICH" && get(ENV, "MPICH_GPU_SUPPORT_ENABLED", "0") == "1"
+            Libdl.dlopen_e(joinpath(dirname(libmpi), "libmpi_gtl_cuda"), Libdl.RTLD_LAZY | Libdl.RTLD_GLOBAL)
```

Comment:
Or rely on the linker finding it automatically just by giving the basename?

Reply:
gtl libs are from a separate rpm, they aren't necessarily in the same directory. You can have multiple `libmpi.so`s (for every compiler, for different version numbers, different fabrics) and just one libmpi_gtl_{cuda,hsa}.so:

```
/opt/cray/pe/mpich/8.1.18/ofi/gnu/9.1/lib/libmpi.so
/opt/cray/pe/mpich/8.1.18/ofi/intel/19.0/lib/libmpi.so
/opt/cray/pe/mpich/8.1.18/ofi/cray/10.0/lib/libmpi.so
/opt/cray/pe/mpich/8.1.18/ofi/nvidia/20.7/lib/libmpi.so
/opt/cray/pe/mpich/8.1.18/ofi/aocc/3.0/lib/libmpi.so
/opt/cray/pe/mpich/8.1.18/ucx/intel/19.0/lib/libmpi.so
/opt/cray/pe/mpich/8.1.18/ucx/cray/10.0/lib/libmpi.so
/opt/cray/pe/mpich/8.1.18/ucx/gnu/9.1/lib/libmpi.so
/opt/cray/pe/mpich/8.1.18/ucx/aocc/3.0/lib/libmpi.so
/opt/cray/pe/mpich/8.1.18/ucx/nvidia/20.7/lib/libmpi.so
/opt/cray/pe/mpich/8.1.21/ofi/gnu/9.1/lib/libmpi.so

...

/opt/cray/pe/mpich/8.1.18/gtl/lib/libmpi_gtl_cuda.so
/opt/cray/pe/mpich/8.1.21/gtl/lib/libmpi_gtl_cuda.so
/opt/cray/pe/mpich/8.1.23/gtl/lib/libmpi_gtl_cuda.so
```

however, there's also a symlink typically for both libmpi.so and the gtl lib in `/opt/cray/pe/lib64`, which is in the ld cache (so no modules required).

Opening by name is probably better than opening by path

Reply:
@simonbyrne see my comments below. To address the questions here:
1. No they are not necessarily in the same place
2. `CC --cray-print-opts=libs` will show which libraries get loaded. This doesn't actually contain any information about _when_ they are needed, but a safe bet would be before `MPI_Init`

Reply:
If they are weak symbols I think they must be loaded before... Otherwise the dlopen on libmpi will try to resolve the weak symbol elsewhere

Reply:
@giordano pointed out that we actually dlopen libmpi in MPIPreferences.__init__, so we may need to do it there:
https://github.com/JuliaParallel/MPI.jl/blob/master/lib/MPIPreferences/src/system.jl#L12

Reply:
I ended up doing: https://github.com/JuliaParallel/MPI.jl/pull/716/files#diff-496cd019dfada6f9862870c26d09d6ea9f6a24aa6d0055fb03e916b8959b82a8R13 -- a list of preloads might be better, because @luraess just told me that other centers might also need cudart preloaded. 😔

Reply:
So on Perlmutter we must `dlopen` before `__init__`. 

But a `Libdl.dlopen_e("libmpi_gtl_cuda", Libdl.RTLD_LAZY | Libdl.RTLD_GLOBAL)` seems to do the trick.